### DISCUSSION TITLE
Docker Backup Workflow?

### DISCUSSION THREAD

**Discussion by disl-pdavid** [ckan/ckan#6984] [Q&A] [answered]
Hi all!

I'm recently elbows-deep in a docker deployment that is working for me so far.
I've bulk loaded a bunch of data into my instance and now I'm planning for the backup/restore workflow but I'm not sure what the best path would be.

Is there a "best practices" workflow for backing up a CKAN instance (specifically with Docker but can be a general workflow that I can adapt)?

At the moment I can brute-force a backup by using `rsync -Pav` on my docker named volumes (the default ones from the docs) but I'm not sure if this is really the best method?

**Comment by duttonw** [comment, marked as answer]
Hi @disl-pdavid ,

Unsure about docker containers, but as a general rule of thumb.

I highly suggest you move your 'resources'/files away from container/network file storage onto a azure/s3/etc blob storage.  IE. https://github.com/qld-gov-au/ckanext-s3filestore/
This then allows your run time web/batch containers to be immutable (replaceable) without dataloss. 

For solr, look over these scripts in this repo https://github.com/qld-gov-au/opswx-ckan-cookbook/blob/master/files/default/solr-sync.sh where we have setup a cron to export out a backup and then blob store it off disk including into a hot solr instance (note that [solrcloud](https://solr.apache.org/guide/6_6/solrcloud.html) could do this but then requires min 6 instances ...)

For database, if you are on a cloud provider, just use AWS RDS or Azure similar product and let that handle those backups and upgrades for you.

If you don't know how to begin on going direct. 

Check out 
https://github.com/qld-gov-au/ckan-qld-infrastructure
To bootstrap with sts role switch (you can just go to ./build-OpenData.sh if you have keys already setup) see below.
note this is running Ansible 2.4.x and has jq app installed

``` #!/bin/bash

set -e

# 1) Assume role for running cloud formation for 3 hours
CREDENTIALS=`aws sts assume-role --role-arn arn:aws:iam::${AWS_RUNTIMEACCOUNT}:role/${AWS_CKAN_ROLE} --external-id ${external id} --role-session-name opsToCkan --duration-seconds 10800 --output=json`
# 2) Capture current credentials to reset after executing command
export ORIGINAL_AWS_ACCESS_KEY_ID=${AWS_ACCESS_KEY_ID}
export ORIGINAL_AWS_SECRET_ACCESS_KEY=${AWS_SECRET_ACCESS_KEY}
# 3) Set AWS Assumed Role Credentials on ENV
export AWS_DEFAULT_REGION=ap-southeast-2
export AWS_ACCESS_KEY_ID=`echo ${CREDENTIALS} | jq -r '.Credentials.AccessKeyId'`
export AWS_SECRET_ACCESS_KEY=`echo ${CREDENTIALS} | jq -r '.Credentials.SecretAccessKey'`
export AWS_SESSION_TOKEN=`echo ${CREDENTIALS} | jq -r '.Credentials.SessionToken'`
export AWS_EXPIRATION=`echo ${CREDENTIALS} | jq -r '.Credentials.Expiration'`
# 4) Execute command as assumed role

aws iam list-account-aliases | grep ${AWS_ALIAS}
greprc=$?
if [ $greprc -eq 0 ]; then
  ./build-CKANTest.sh
# OR   ./build-OpenData.sh
else 
  echo "we are not in ckan account, we did not run script";
  exit 1;
fi

# 5) Clean up assume role environment variables
unset AWS_ACCESS_KEY_ID
unset AWS_SECRET_ACCESS_KEY
unset AWS_SESSION_TOKEN
unset AWS_EXPIRATION
# 6) Reset to original credentials
export AWS_ACCESS_KEY_ID=${ORIGINAL_AWS_ACCESS_KEY_ID}
export AWS_SECRET_ACCESS_KEY=${ORIGINAL_AWS_SECRET_ACCESS_KEY}
# 7) Clean up temporary storage
unset ORIGINAL_AWS_ACCESS_KEY_ID
unset ORIGINAL_AWS_SECRET_ACCESS_KEY 
```

This is running www.data.qld.gov.au as well as www.publications.qld.gov.au 

Regards,

@duttonw

**Reply by disl-pdavid** [reply]
Fantastic write-up, thank you @duttonw!  Much appreciated.

I am at a good proof-of-concept for my stakeholders and just want to have a backup of the existing site while I get approval for a move onto more robust services (AWS/Google Cloud stuff eventually).  This helps immensely for decisions that'll have to be made soon. :)

Thank you!

**Comment by ghost** [comment]
Not sure what you're docker setup is, but if you mount docker volumes into the different containers (CKAN, SOLR, Postgres), you can use [this image](https://github.com/offen/docker-volume-backup) to create backups of these volumes to many different locations. The repository also contains an explanation on how to restore the backup.
